From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 15 Nov 2025 01:50:25 +0200
Subject: [PATCH] Respect ghost recipes in PlayerRecipeBookClickEvent

Skipped: Was made for a feature that didn't quite work out, and as is, clicking for the second time does not fire the PlayerRecipeBookClickEvent for whatever reason


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index e30807eb7bb9a78e7b0616166d61c9aed4fa4bfa..e6e9c4d4e3b232ce2c30c6e21f6cc4f6e5ac7789 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -3408,7 +3408,18 @@ public class ServerGamePacketListenerImpl
                             );
                             // CraftBukkit end
                             if (postPlaceAction == RecipeBookMenu.PostPlaceAction.PLACE_GHOST_RECIPE) {
-                                this.send(new ClientboundPlaceGhostRecipePacket(this.player.containerMenu.containerId, recipeFromDisplay.display().display()));
+                                // Kiterino start - Respect ghost recipes in PlayerRecipeBookClickEvent
+                                if (!this.player.getRecipeBook().contains(recipeHolder.id())) {
+                                    return;
+                                }
+                                List<net.minecraft.world.item.crafting.display.RecipeDisplay> list = new java.util.ArrayList<>();
+                                this.player.getRecipeBook().displayResolver.displaysForRecipe(recipeHolder.id(), entry -> list.add(entry.display()));
+                                if (list.isEmpty()) {
+                                    return;
+                                }
+                                net.minecraft.world.item.crafting.display.RecipeDisplay recipeDisplay = list.getFirst();
+                                this.send(new ClientboundPlaceGhostRecipePacket(this.player.containerMenu.containerId, recipeDisplay));
+                                // Kiterino end - Respect ghost recipes in PlayerRecipeBookClickEvent
                             }
                         }
                     }
